Some docs for workloadmeta

djmitche opened the pull request:
### What does this PR do?

Adds docs for workloadmeta

### Motivation

Hoping to save the next person some of the head-scratching I just did :)

### Additional Notes

I know this is just docs and it's easy to say "sure looks fine", but please verify my statements as I may be wrong or misunderstand the situation.

Particular questions = places where I'm not sure:
 * Is my description of the first EventBundle correct?
 * Is my description of the peculiarities of unset events correct?
   * If so, how can a subscriber know when an entity is really "gone"?
   * [These two conditionals](https://github.com/DataDog/datadog-agent/blob/f4cbdd633850263e8aab363c008087ab8b2ed72d/pkg/workloadmeta/store.go#L468-L483) have the same body.  Was one of those supposed to be different?  Perhaps when it's not gone entirely, the Sources and Entity should be merged from the remaining sources?
 * I suggested always checking the cast of an Entity to a concrete type, but I noticed that's rarely done.  Are there more stringent limits on when an Entity might not have a concrete type corresponding to entity.GetID().Kind, that mean it's OK to do an unchecked cast outside those limits?
 * Why would a subscriber choose to filter events from only one source?

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

juliogreff committed:
[workloadmeta] Remove Event's Sources

This is currently unused, and following commits will remove the
capability of calling code to pass in filters with multiple sources
anyway, because this ended up being a misfeature anyway.

juliogreff committed:
[workloadmeta] Refactor source filtering

Previously, each collector pretty much owned its own `Source` (as in,
the `kubelet` collector used a `kubelet` source, the `docker` collector
used a `docker` source, and so on). This implies that components that
filter by container runtimes would need to update their filters for
every new container runtime the workloadmeta supports, which is less
than ideal.

To solve that, the sources have been reworked. Now we have the following
sources:

* SourceRuntime: entities detected by the container runtime running on
  the node, collecting lower level information about containers.
  `docker`, `containerd`, `podman` and `ecs_fargate` use this source.
* SourceNodeOrchestrator: entities detected by the node agent from an
  orchestrator. `kubelet` and `ecs` use this.
* SourceClusterOrchestrator: entities detected by calling the central
  component of an orchestrator, or the Datadog Cluster Agent.
  `kube_metadata` and `cloudfoundry` use this.

This commit also removes the ability to filter by more than one source,
as that's no longer necessary, simplifying the code a lot.

juliogreff committed:
[workloadmeta] Introduce SourceAll instead of magic empty string

juliogreff committed:
[AD] Fargate uses SourceRuntime as that's what workloadmeta collects

djmitche committed:
Improve WLM documentation

djmitche commented:
BTW, I'm not done here, but I'm off until Monday so I thought I'd get this wrapped up into a PR while it's still hot in my memory.  I'll make more PRs to add more docs / clarity.

davidor commented on pkg/workloadmeta/types.go:
```diff
@@ -76,46 +76,71 @@ const (
 	Merge(Entity) error
+
+	// DeepCopy copies an entity such that modifications of the copy will not
+	// affet the original.
```
typo `affet`.

davidor commented on pkg/workloadmeta/README.md:
```diff
@@ -0,0 +1,32 @@
+# Workloadmeta Store
```
It might be worth to mention somewhere in this doc that we have telemetry for the workloadmeta store: https://github.com/DataDog/datadog-agent/blob/main/pkg/workloadmeta/telemetry/telemetry.go

davidor replied:
We also have the `agent workload-list` command. It's useful for debugging. It might be good to mention it somewhere.

djmitche commented on cmd/agent/common/loader.go:
```diff
@@ -30,6 +30,9 @@ func LoadComponents(confdPath string) {
 		workloadmeta.GetGlobalStore().Start(context.Background())
 
+		// uncomment this to dump all workloadmeta events as debug logs
+		//dumper.Enable()
```
This was intended as a debugging tool.  It helped me see some events, and I was imagining myself adding a collector some new service and wanting to compile this in so I could eyeball the events my new code was generating.  So, more than telemetry, and more than `workload-list`.

So it's not something that would be enabled on a running agent, and I put it in a different package to avoid adding package size.  I'd be happy to remove this entirely -- I figured since I'd written it I might as well see if we wanted to keep it :)

djmitche commented on pkg/workloadmeta/README.md:
```diff
@@ -0,0 +1,32 @@
+The current workload of a host or cluster is represented by the current set of entities.
+
+Each _Entity_ has a unique _EntityID_, composed of a _Kind_ and an ID.
+Supported kinds include container, pod, and task.
```
I wanted to avoid trying to enumerate everything that's supported, since inevitably such a list falls out of date.  I'll update the text to say "Examples of kinds include .." so that it's a bit clearer that the list is illustrative and not exhaustive.

juliogreff commented on pkg/workloadmeta/types.go:
```diff
@@ -546,16 +555,40 @@ type CollectorEvent struct {
-	Entity  Entity
+
+	// Entity is the entity involved in this event.  For a set event, this may
+	// contain information "merged" from multiple sources, as listed in the
```
#10331 removes `Sources` entirely, as we want to discourage clients from trying to guess what's generating events (this whole thing is supposed to be source-agnostic after all), so I wouldn't even mention merging here, as it's just an internal implementation detail.

djmitche replied:
Done.

juliogreff commented on pkg/workloadmeta/types.go:
```diff
@@ -546,16 +555,40 @@ type CollectorEvent struct {
+	// When Type is EventTypeSet, this represents an added or updated entity.
+	// Multiple set events may be sent for a single entity.
+	//
+	// When Type is EventTypeUnset, this represents either a removed entity or
```
When a source removes the information for an entity, we decide whether or not to send a Set or Unset based on filters. If the subscriber has a filter, and it filters out the info from the remaining sources, we send an `Unset` (as from that subscriber's point of view, their entity is gone). If there are remaining sources, it's just a simple update, so we send a `Set`. So it's not possible for a subscriber to receive more than one Unset per entity (unless my code's broken, of course 😅).

I agree it's very hard to make that out from reading the code, and that's one reason why #10331 changes this significantly.

djmitche replied:
Ah, should this 
https://github.com/DataDog/datadog-agent/blob/e8bae0f8d181e9e84653a33f259dd5135774a16f/pkg/workloadmeta/store.go#L435-L438
possibly be EventTypeSet, then?  That answers one of my questions :)

juliogreff commented on pkg/workloadmeta/types.go:
```diff
@@ -80,55 +63,84 @@ const (
+	// EventTypeSet indicates that an entity has been added or updated.
 	EventTypeSet EventType = iota
+
+	// EventTypeUnset indicates that an entity has been removed, or that one
+	// source of several sources provided information about an entity has
+	// stopped providing such information.
```
Explained elsewhere in the thread that this is not accurate, Unset really means an entity has been removed.

juliogreff commented on pkg/workloadmeta/store.go:
```diff
@@ -396,29 +440,29 @@ func (s *store) handleEvents(evs []CollectorEvent) {
 		}
 
-		entityOfSource, ok := entitiesOfKind[meta.ID]
+		entitiesBySource, ok := entitiesOfKind[meta.ID]
```
This is a good rename, but this has also been changed in #10331. It would save us both a lot of work to just revert and avoid the conflicts.

djmitche replied:
I'll drop that commit when I rebase onto #10331.

juliogreff commented on pkg/workloadmeta/store.go:
```diff
@@ -30,6 +30,50 @@ const (
+	// store, limited to events matching the filter.  The name is used for
+	// telemetry and debugging.
+	//
+	// The first message on the channel is special: it contains an EventTypeSet
+	// event for each entity currently in the store.  If the Subscribe call
+	// occurs during agent startup, then it is safe to assume that entities
+	// referenced in the first message were running before the agent started.
```
Might be worth point it out that this is inherently racy, so assuming that anything NOT in the first message was NOT running before the agent started could very well be wrong ;)

djmitche replied:
Good point, thanks.

juliogreff commented on pkg/workloadmeta/store.go:
```diff
@@ -30,6 +30,50 @@ const (
+	// occurs during agent startup, then it is safe to assume that entities
+	// referenced in the first message were running before the agent started.
+	//
+	// When more than one source provides information about an entity in a
+	// EventTypeSet event, the information from all sources included by the
+	// filter will be merged into one entity, and the event's Sources property
+	// will contain the names of all associated sources.
```
See comment elsewhere about `Sources` being removed.

juliogreff commented on pkg/workloadmeta/store.go:
```diff
@@ -30,6 +30,50 @@ const (
+	// filter will be merged into one entity, and the event's Sources property
+	// will contain the names of all associated sources.
+	//
+	// EventTypeUnset events contain only information from a single source, and
+	// the entity contains only the information provided by that source.  An unset
+	// event is sent as each source removes its information for an entity, so an
+	// entity with multiple sources may result in multiple unset events.
```
... and also how there can only be one unset.

juliogreff reviewed (commented):
Added comments where appropriate, overall good stuff, thanks for taking the time! You've already been made aware of #10331, but I don't know who should merge what first. I don't intend to merge that PR for 7.34 (since 7.33 had a ton of related changes already, and it's not released yet, so trying not to rock the boat too hard all at once), so I'd be ok merging this now, but on the other hand it's just docs, and as far as I'm aware we're the only people working in this area. Happy to hear your thoughts.

Answers to your other questions:

> I suggested always checking the cast of an Entity to a concrete type, but I noticed that's rarely done. Are there more stringent limits on when an Entity might not have a concrete type corresponding to entity.GetID().Kind, that mean it's OK to do an unchecked cast outside those limits?

On an `EventTypeSet`, it's always ok to do an unchecked cast, as it makes no sense to emit an entity without any information, and lying about its kind would clearly be a bug that I'd prefer to blow up in my face ASAP. 

`EventTypeUnset` _currently_ just emits an `EntityID`, but I'm changing it soon to include more information about the last moments of an entity's lifetime (details TBD). Once that's the done it would be OK to do unchecked casts for everything.

> Why would a subscriber choose to filter events from only one source?

We have one particular use case, which is to report container lifecycle events. Filtering by only container runtime events, we limit the amount of events that subscriber would have to process, and get an `EventTypeUnset` slightly sooner.

djmitche commented on pkg/workloadmeta/dumper/dumper.go:
```diff
@@ -0,0 +1,61 @@
+		close(eventBundle.Ch)
+
+		for i, evt := range eventBundle.Events {
+			prefix := fmt.Sprintf("WLM %d/%d", bundleCount, i)
```
Done.

djmitche commented:
> I don't know who should merge what first

I do -- #10331.  I'll rebase this PR on top of that one, and land after it.  The primary goal here is for me to understand things; updating the docs is only secondary and there's no rush.

> EventTypeUnset currently just emits an EntityID, but I'm changing it soon to include more information about the last moments of an entity's lifetime (details TBD). Once that's the done it would be OK to do unchecked casts for everything.

OK, I'll document it as it is now, and we can adjust the docs as necessary.

So, I think the remaining bit I'd like you to look at right now (aside from making sure I've translated your comments correctly) is the Unset logic, mentioned in a reply to your comment.

djmitche commented:
I added one more commit documenting Filter, too -- but it's probably not very controversial.

djmitche commented:
Ah, I see that bug is fixed in https://github.com/DataDog/datadog-agent/pull/10513.

juliogreff committed:
[workloadmeta] Fix merge conflicts

djmitche commented:
I'm going to drop the dumper support -- I think that was useful enough to me to develop this, but not worth keeping long-term.

djmitche committed:
additional docs on top of latest #10331

djmitche commented:
OK, rebased on top of #10331 again.  I squashed all of the previous commits, but I've included today's changes in a new commit for your reviewing pleasure :)

djmitche commented:
Whoops, that rebase was to an old commit.  One moment.

djmitche commented:
Fixed :)

djmitche commented:
@juliogreff if you can take a look at the last commit, I think that will be enough to merge this.

juliogreff commented on pkg/workloadmeta/types.go:
```diff
@@ -618,15 +635,14 @@ type Event struct {
-	// perhaps only an EntityID.
+	// Entity is the entity involved in this event.  For an EventTypeSet event,
+	// this may contain information "merged" from multiple sources, as listed
+	// in the Sources field.  For an unset event it contains only an EntityID.
```
Everything is correct except the `Sources` field is no more.

djmitche replied:
Good catch, thanks!

juliogreff reviewed (approved):
Super minor nit, but I think we're basically good to go. Sorry it took so long to get this merged 🙏

djmitche committed:
remove ref to now-removed Sources field

djmitche committed:
Merge branch 'main' into dustin.mitchell/wlm-polish
